src/providers/hana/qgshanaprovider.cpp

```diff
@@ -0,0 +1,1780 @@
+  return static_cast<long>( count );
+}
+
+QgsCoordinateReferenceSystem QgsHanaProvider::crs() const
+{
+  QgsCoordinateReferenceSystem srs;
+
+  static QMutex sMutex;
+  QMutexLocker locker( &sMutex );
+  static QMap<int, QgsCoordinateReferenceSystem> sCrsCache;
+  if ( sCrsCache.contains( mSrid ) )
+  {
+    srs = sCrsCache.value( mSrid );
+    return srs;
+  }
+
+  QgsHanaConnectionRef conn( mUri );
+  srs = conn->getCrs( mSrid );
+
+  if ( srs.isValid() )
+    sCrsCache.insert( mSrid, srs );
+
+  return srs;
+}
+
+QgsVectorLayerExporter::ExportError QgsHanaProvider::createEmptyLayer(
+  const QString &uri,
+  const QgsFields &fields,
+  QgsWkbTypes::Type wkbType,
+  const QgsCoordinateReferenceSystem &srs,
+  bool overwrite,
+  QMap<int, int> *oldToNewAttrIdxMap,
+  QString *errorMessage,
+  const QMap<QString, QVariant> *
+)
+{
+  QgsDataSourceUri dsUri( uri );
+  QgsHanaConnectionRef conn( dsUri );
+  if ( conn.isNull() )
+  {
+    if ( errorMessage )
+      *errorMessage = QObject::tr( "Connection to database failed" );
+    return QgsVectorLayerExporter::ErrConnectionFailed;
+  }
+
+  QString schemaName = dsUri.schema();
+  QString tableName = dsUri.table();
+
+  if ( schemaName.isEmpty() )
+  {
+    if ( errorMessage )
+      *errorMessage = QObject::tr( "Schema name cannot be empty" );
+    return QgsVectorLayerExporter::ErrCreateLayer;
+  }
+
+  QString geometryColumn = dsUri.geometryColumn();
+  QString geometryType;
+
+  QString primaryKey = dsUri.keyColumn();
+  QString primaryKeyType;
+
+  QString schemaTableName = QgsHanaUtils::quotedIdentifier( schemaName ) + '.' +
+                            QgsHanaUtils::quotedIdentifier( tableName );
+
+  if ( wkbType != QgsWkbTypes::NoGeometry && geometryColumn.isEmpty() )
+    geometryColumn = "geom";
+
+  bool fieldsInUpperCase = false;
+  if ( fields.size() > 0 )
+  {
+    int count = QgsHanaUtils::countFieldsWithFirstLetterInUppercase( fields );
+    fieldsInUpperCase = count > fields.size() / 2;
+  }
+
+  bool createdNewPk = false;
+
+  if ( primaryKey.isEmpty() )
+  {
+    QString pk = primaryKey = fieldsInUpperCase ? "ID" : "id";
+    int index = 0;
+    while ( fields.indexFromName( primaryKey ) >= 0 )
+    {
+      primaryKey = QStringLiteral( "%1_%2" ).arg( pk ).arg( index++ );
+    }
+
+    createdNewPk = true;
+  }
+  else
+  {
+    int idx = fields.indexFromName( primaryKey );
+    if ( idx >= 0 )
+    {
+      QgsField fld = fields.at( idx );
+      if ( QgsHanaUtils::convertField( fld ) )
+        primaryKeyType = fld.typeName();
+    }
+  }
+
+  if ( primaryKeyType.isEmpty() )
+    primaryKeyType = QStringLiteral( "BIGINT" );
+
+  QString sql;
+
+  // set up spatial reference id
+  long srid = 0;
+  if ( srs.isValid() )
+  {
+    srid = srs.postgisSrid();
+    QString authSrid = QStringLiteral( "null" );
+    QString authName = QStringLiteral( "null" );
+    QStringList sl = srs.authid().split( ':' );
+    if ( sl.length() == 2 )
+    {
+      authName = '\'' + sl[0] + '\'';
```

Comment:
As we use a prepared statement below, there shouldn't be any quotes.

Reply:
Done in 5ce5c03.